File: claude_setup/domain/core_kp_routing.py
from __future__ import annotations

from dataclasses import dataclass
from typing import List

from claude_setup.models import ProjectConfig


@dataclass(frozen=True)
class CoreKpRoute:
    """Maps a core rule source file to a knowledge pack destination."""

    source_file: str
    kp_name: str
    dest_file: str


@dataclass(frozen=True)
class ConditionalCoreKpRoute(CoreKpRoute):
    """Route that is conditionally included based on config values."""

    condition_field: str = ""
    condition_exclude: str = ""


CORE_TO_KP_MAPPING: List[CoreKpRoute] = [
    CoreKpRoute("01-clean-code.md", "coding-standards", "clean-code.md"),
    CoreKpRoute("02-solid-principles.md", "coding-standards", "solid-principles.md"),
    CoreKpRoute("03-testing-philosophy.md", "testing", "testing-philosophy.md"),
    CoreKpRoute("05-architecture-principles.md", "architecture", "architecture-principles.md"),
    CoreKpRoute("06-api-design-principles.md", "api-design", "api-design-principles.md"),
    CoreKpRoute("07-security-principles.md", "security", "security-principles.md"),
    CoreKpRoute("08-observability-principles.md", "observability", "observability-principles.md"),
    CoreKpRoute("09-resilience-principles.md", "resilience", "resilience-principles.md"),
    CoreKpRoute("10-infrastructure-principles.md", "infrastructure", "infrastructure-principles.md"),
    CoreKpRoute("11-database-principles.md", "database-patterns", "database-principles.md"),
    CoreKpRoute("13-story-decomposition.md", "story-planning", "story-decomposition.md"),
]

CONDITIONAL_CORE_KP: List[ConditionalCoreKpRoute] = [
    ConditionalCoreKpRoute(
        source_file="12-cloud-native-principles.md",
        kp_name="infrastructure",
        dest_file="cloud-native-principles.md",
        condition_field="architecture_style",
        condition_exclude="library",
    ),
]


def get_active_routes(config: ProjectConfig) -> List[CoreKpRoute]:
    """Return all routes whose conditions are met for the given config."""
    routes: List[CoreKpRoute] = list(CORE_TO_KP_MAPPING)
    for route in CONDITIONAL_CORE_KP:
        config_value = _resolve_condition_value(config, route.condition_field)
        if config_value != route.condition_exclude:
            routes.append(route)
    return routes


def _resolve_condition_value(config: ProjectConfig, field: str) -> str:
    """Resolve a dotted condition field to a config value."""
    if field == "architecture_style":
        return config.architecture.style
    return ""

File: src/domain/core-kp-routing.ts
/**
 * Core rule to knowledge pack routing.
 *
 * Migrated from Python `domain/core_kp_routing.py`.
 */

import type { ProjectConfig } from "../models.js";

/** Maps a core rule source file to a knowledge pack destination. */
export interface CoreKpRoute {
  readonly sourceFile: string;
  readonly kpName: string;
  readonly destFile: string;
}

/** Supported condition fields for conditional routing. */
export type ConditionField = "architecture_style";

/** Route that is conditionally included based on config values. */
export interface ConditionalCoreKpRoute extends CoreKpRoute {
  readonly conditionField: ConditionField;
  readonly conditionExclude: string;
}

/** 12 static routes from core rules to knowledge packs. */
export const CORE_TO_KP_MAPPING: readonly CoreKpRoute[] = [
  { sourceFile: "01-clean-code.md", kpName: "coding-standards", destFile: "clean-code.md" },
  { sourceFile: "02-solid-principles.md", kpName: "coding-standards", destFile: "solid-principles.md" },
  { sourceFile: "03-testing-philosophy.md", kpName: "testing", destFile: "testing-philosophy.md" },
  { sourceFile: "05-architecture-principles.md", kpName: "architecture", destFile: "architecture-principles.md" },
  { sourceFile: "06-api-design-principles.md", kpName: "api-design", destFile: "api-design-principles.md" },
  { sourceFile: "07-security-principles.md", kpName: "security", destFile: "security-principles.md" },
  { sourceFile: "08-observability-principles.md", kpName: "observability", destFile: "observability-principles.md" },
  { sourceFile: "09-resilience-principles.md", kpName: "resilience", destFile: "resilience-principles.md" },
  { sourceFile: "10-infrastructure-principles.md", kpName: "infrastructure", destFile: "infrastructure-principles.md" },
  { sourceFile: "11-database-principles.md", kpName: "database-patterns", destFile: "database-principles.md" },
  { sourceFile: "13-story-decomposition.md", kpName: "story-planning", destFile: "story-decomposition.md" },
  { sourceFile: "14-refactoring-guidelines.md", kpName: "coding-standards", destFile: "refactoring-guidelines.md" },
];

/** 1 conditional route: cloud-native excluded for library style. */
export const CONDITIONAL_CORE_KP: readonly ConditionalCoreKpRoute[] = [
  {
    sourceFile: "12-cloud-native-principles.md",
    kpName: "infrastructure",
    destFile: "cloud-native-principles.md",
    conditionField: "architecture_style",
    conditionExclude: "library",
  },
];

/** Return all routes whose conditions are met for the given config. */
export function getActiveRoutes(config: ProjectConfig): CoreKpRoute[] {
  const routes: CoreKpRoute[] = [...CORE_TO_KP_MAPPING];
  for (const route of CONDITIONAL_CORE_KP) {
    const configValue = resolveConditionValue(config, route.conditionField);
    if (configValue !== route.conditionExclude) {
      routes.push(route);
    }
  }
  return routes;
}

/** Resolve a condition field name to a config value. */
function resolveConditionValue(config: ProjectConfig, field: ConditionField): string {
  switch (field) {
    case "architecture_style":
      return config.architecture.style;
  }
}
